Clear Inline Functions

TheJaredWilcurt opened the pull request:
[![Coverage Status](https://coveralls.io/repos/github/tjw-lint/jest-serializer-vue-tjw/badge.svg?branch=clear-fn)](https://coveralls.io/github/tjw-lint/jest-serializer-vue-tjw?branch=clear-fn)

* [x] Functionality
* [x] Tests
* [x] Documentation
* [x] Version bump
* [x] ~~Update deps~~

TheJaredWilcurt committed:
First draft

TheJaredWilcurt commented on src/stringManipulation.js:
```diff
@@ -93,6 +93,34 @@ function clearAttributes (html, options) {
+    $('*').each(function (index, element) {
+      Object.keys(element.attribs).forEach(function (attributeName) {
+        let value = element.attribs[attributeName].trim();
+        if (value.startsWith('function') && value.endsWith('}')) {
```
this `if` statement needs replaced with a more robust regex pattern

TheJaredWilcurt replied:
```suggestion
        let pattern = /function *\(([a-zA-Z0-9_$],* *){0,}\) *{(.*\n*)*}/;
        if (pattern.test(value)) {
```

TheJaredWilcurt committed:
Add regex pattern

TheJaredWilcurt committed:
bump

TheJaredWilcurt committed:
bump

cameron-martin commented:
There's a bunch of cases where this will break and I don't think this is the right approach personally, but otherwise it gets the job done in most cases I guess.

TheJaredWilcurt commented:
@cameron-martin can you elaborate with examples. I have never used inline functions, so I'm trying to create this feature for those that do. That's why I'm asking for your input.

cameron-martin commented:
Cases where this will break that I can see are:

* Where the prop is actually a string that happens to contain something that matches the regex.
* Pretty much all the cases where you do ` *` need to take into account other whitespace.
* Identifiers in JS can include way more than just `[a-zA-Z0-9_$]`.
* Unless I'm missing something, it assumes transpilation of arrow functions, since the regex doesn't match arrow functions. This may not be the case.

The last ones are fixable by changing the regex, but the first one isn't, although may not be much of an issue in practice.

TheJaredWilcurt commented on tests/__snapshots__/InlineFunctions.test.js.snap:
```diff
@@ -0,0 +1,29 @@
+      }">Multi-line</h2>
+  <h3 title="function (x) {
+        return !x;
+      }">Arrow</h3>
```
@cameron-martin You can see here what Vue renders the templates out to. the One liner, Multi-line, and Arrow function, are all output as the same multi-line approach.

cameron-martin replied:
Interesting. How does that transformation get peformed?

TheJaredWilcurt replied:
Not sure when or how, but by the time we are checking for it, they are all the same.

TheJaredWilcurt committed:
Improve tests and pattern match

TheJaredWilcurt commented:
@cameron-martin 

>  Where the prop is actually a string that happens to contain something that matches the regex.

`<div title="function stuff () { whatever; }">` yes, this would get converted to `<div title="[function]">`, but this is an opt-in feature. You need to set `clearInlineFunctions: true` in the API. So this should catch anyone off-guard.

> Pretty much all the cases where you do * need to take into account other whitespace.
> Identifiers in JS can include way more than just [a-zA-Z0-9_$].

I've updated the regex and added in test cases for it. Though I am very open to PR suggestions for improvements to the regex or tests around the regex.

> Unless I'm missing something, it assumes transpilation of arrow functions, since the regex doesn't match arrow functions. This may not be the case.

Yes, Vue converts all inline functions to look the same by the time the regex checks it.

cameron-martin commented:
Looks good to me, then.

TheJaredWilcurt committed:
strip down string manip

TheJaredWilcurt committed:
Create cheerio manip

TheJaredWilcurt committed:
Update deps
